File: Sources/BraintreeVenmo/BTVenmoAppSwitchReturnURL.m
#import "BTVenmoAppSwitchReturnURL.h"

// Swift Module Imports
#if __has_include(<Braintree/Braintree-Swift.h>) // CocoaPods
#import <Braintree/Braintree-Swift.h>

#elif SWIFT_PACKAGE                              // SPM
/* Use @import for SPM support
 * See https://forums.swift.org/t/using-a-swift-package-in-a-mixed-swift-and-objective-c-project/27348
 */
@import BraintreeCore;

#elif __has_include("Braintree-Swift.h")         // CocoaPods for ReactNative
/* Use quoted style when importing Swift headers for ReactNative support
 * See https://github.com/braintree/braintree_ios/issues/671
 */
#import "Braintree-Swift.h"

#else                                            // Carthage
#import <BraintreeCore/BraintreeCore-Swift.h>
#endif

NSString *const BTVenmoAppSwitchReturnURLErrorDomain = @"com.braintreepayments.BTVenmoAppSwitchReturnURLErrorDomain";

@implementation BTVenmoAppSwitchReturnURL

+ (BOOL)isValidURL:(NSURL *)url {
    return [url.host isEqualToString:@"x-callback-url"] && [url.path hasPrefix:@"/vzero/auth/venmo/"];
}

- (instancetype)initWithURL:(NSURL *)url {
    self = [self init];
    if (self) {
        NSDictionary *parameters = [BTURLUtils queryParametersForURL:url];
        if ([url.path isEqualToString:@"/vzero/auth/venmo/success"]) {
            if (parameters[@"resource_id"]) {
                _state = BTVenmoAppSwitchReturnURLStateSucceededWithPaymentContext;
                _paymentContextID = parameters[@"resource_id"];
            } else {
                _state = BTVenmoAppSwitchReturnURLStateSucceeded;
                _nonce = parameters[@"paymentMethodNonce"];
                _username = parameters[@"username"];
            }
        } else if ([url.path isEqualToString:@"/vzero/auth/venmo/error"]) {
            _state = BTVenmoAppSwitchReturnURLStateFailed;
            NSString *errorMessage = parameters[@"errorMessage"];
            NSInteger errorCode = [parameters[@"errorCode"] integerValue];
            _error = [NSError errorWithDomain:BTVenmoAppSwitchReturnURLErrorDomain code:errorCode userInfo:(errorMessage != nil ? @{ NSLocalizedDescriptionKey: errorMessage } : nil)];
        } else if ([url.path isEqualToString:@"/vzero/auth/venmo/cancel"]) {
            _state = BTVenmoAppSwitchReturnURLStateCanceled;
        } else {
            _state = BTVenmoAppSwitchReturnURLStateUnknown;
        }
    }
    return self;
}

@end

File: Sources/BraintreeVenmo/BTVenmoAppSwitchReturnURL.swift
import Foundation

#if canImport(BraintreeCore)
import BraintreeCore
#endif

enum BTVenmoAppSwitchReturnURLState {
    case unknown
    case succeededWithPaymentContext
    case succeeded
    case failed
    case canceled
}

///  This class interprets URLs received from the Venmo app via app switch returns.
///
///  Venmo Touch app switch authorization requests should result in success, failure or user-initiated cancelation. These states are communicated in the url.
struct BTVenmoAppSwitchReturnURL {

    // MARK: - Properties

    /// The overall status of the app switch - success, failure or cancelation
    var state: BTVenmoAppSwitchReturnURLState = .unknown

    /// The nonce from the return URL.
    var nonce: String?

    /// The username from the return URL.
    var username: String?

    /// The payment context ID from the return URL.
    var paymentContextID: String?

    /// If the return URL's state is BTVenmoAppSwitchReturnURLStateFailed, the error returned from Venmo via the app switch.
    var error: Error?

    // MARK: - Initializer

    /// Initializes a new BTVenmoAppSwitchReturnURL
    /// - Parameter url: an incoming app switch url
    init?(url: URL) {
        let parameters = BTURLUtils.queryParameters(for: url)

        if url.path.contains("success") {
            if let resourceID = parameters["resource_id"] {
                state = .succeededWithPaymentContext
                paymentContextID = resourceID
            } else {
                state = .succeeded
                nonce = parameters["paymentMethodNonce"] ?? parameters["payment_method_nonce"]
                username = parameters["username"]
            }
        } else if url.path.contains("error") {
            state = .failed
            let errorMessage: String? = parameters["errorMessage"] ?? parameters["error_message"]
            let errorCode = Int(parameters["errorCode"] ?? parameters["error_code"] ?? "0")
            error = BTVenmoAppSwitchError.returnURLError(errorCode ?? 0, errorMessage)
        } else if url.path.contains("cancel") {
            state = .canceled
        } else {
            state = .unknown
        }
    }

    // MARK: - Internal Methods

    /// Evaluates whether the url represents a valid Venmo return.
    /// - Parameter url: an app switch return URL
    /// - Returns: `true` if the url represents a Venmo app switch return
    static func isValid(url: URL) -> Bool {
        /// universal link path components
        let isHTTPSScheme = url.scheme == "https"
        let containsAppSwitchPath = url.path.contains("braintreeAppSwitchVenmo")
        let containsExpectedPath = url.path.contains("cancel") || url.path.contains("success") || url.path.contains("error")
        let isValidAppSwitchURL = isHTTPSScheme && containsAppSwitchPath && containsExpectedPath
        
        /// url scheme expected host and path components
        let isValidURLSchemeURL = url.host == "x-callback-url" && url.path.hasPrefix("/vzero/auth/venmo/")

        return isValidURLSchemeURL || isValidAppSwitchURL
    }
}
